packages/squarev2-integration/src/squarev2-payment-strategy.ts

```diff
@@ -142,19 +104,54 @@ export default class SquareV2PaymentStrategy implements PaymentStrategy {
         return features['PROJECT-3828.add_3ds_support_on_squarev2'];
     }
 
-    private async _processWithVaultedInstrument(
-        payment: OrderPaymentRequestBody,
-        shouldSaveInstrument = false,
-        shouldSetAsDefaultInstrument = false,
-    ): Promise<void> {
-        const { methodId, paymentData } = payment;
+    private async _getCardPayload(shouldSaveInstrument?: boolean) {
+        const cardTokenizationResult = await this._squareV2PaymentProcessor.tokenize();
+
+        if (!this._shouldVerify()) {
+            return {
+                credit_card_token: {
+                    token: cardTokenizationResult,
+                },
+            };
+        }
 
-        // TODO: optimize type guard (the same in execute method)
-        // need some refactoring in execute method
-        if (!paymentData || !isVaultedInstrument(paymentData)) {
-            return;
+        const creditCardTokens = {
+            nonce: cardTokenizationResult,
+            token: await this._squareV2PaymentProcessor.verifyBuyer(
+                cardTokenizationResult,
+                SquareIntent.CHARGE,
+            ),
+        };
+        let savingCreditCardTokens;
+
+        if (shouldSaveInstrument) {
+            // INFO: additional 'tokenize' is required to verify and save the card
+            // for each 'verifyBuyer' we need to generate new token
+            const tokenForSavingCard = await this._squareV2PaymentProcessor.tokenize();
+
+            savingCreditCardTokens = {
+                store_card_nonce: tokenForSavingCard,
+                store_card_token: await this._squareV2PaymentProcessor.verifyBuyer(
+                    tokenForSavingCard,
+                    SquareIntent.STORE,
+                ),
+            };
```

Comment:
It looks good to me, but I wan to suggest another variant:

```suggestion
        let tokenData = {
            nonce: cardTokenizationResult,
            token: await this._squareV2PaymentProcessor.verifyBuyer(
                cardTokenizationResult,
                SquareIntent.CHARGE,
            ),
        };

        if (shouldSaveInstrument) {
            // INFO: additional 'tokenize' is required to verify and save the card
            // for each 'verifyBuyer' we need to generate new token
            const tokenForSavingCard = await this._squareV2PaymentProcessor.tokenize();

            tokenData = {
                ...tokenData,
                store_card_nonce: tokenForSavingCard,
                store_card_token: await this._squareV2PaymentProcessor.verifyBuyer(
                    tokenForSavingCard,
                    SquareIntent.STORE,
                ),
            };
```

Reply:
I will apply your variant, thanks!